src/gfx/convert.cpp

```diff
@@ -0,0 +1,831 @@
+
+		fatal("Error reading input image (\"%s\"): %s", self->path.c_str(), msg);
+	}
+
+	static void handleWarning(png_structp png, char const *msg) {
+		struct Png *self = reinterpret_cast<Png *>(png_get_error_ptr(png));
+
+		warning("In input image (\"%s\"): %s", self->path.c_str(), msg);
+	}
+
+	static void readData(png_structp png, png_bytep data, size_t length) {
+		struct Png *self = reinterpret_cast<Png *>(png_get_io_ptr(png));
+		std::streamsize expectedLen = length;
+		std::streamsize nbBytesRead = self->file.sgetn(reinterpret_cast<char *>(data), expectedLen);
+
+		if (nbBytesRead != expectedLen) {
+			fatal("Error reading input image (\"%s\"): file too short (expected at least %zd more "
+			      "bytes after reading %lld)",
+			      self->path.c_str(), length - nbBytesRead,
+			      self->file.pubseekoff(0, std::ios_base::cur));
+		}
+	}
+
+public:
+	ImagePalette const &getColors() const { return colors; }
+
+	int getColorType() const { return colorType; }
+
+	std::tuple<int, png_const_colorp, png_bytep> getEmbeddedPal() const {
+		return {nbColors, embeddedPal, transparencyPal};
+	}
+
+	uint32_t getWidth() const { return width; }
+
+	uint32_t getHeight() const { return height; }
+
+	uint16_t &pixel(uint32_t x, uint32_t y) { return pixels[y * width + x]; }
+
+	uint16_t const &pixel(uint32_t x, uint32_t y) const { return pixels[y * width + x]; }
+
+	bool hasNonGray() const { return !isGrayOnly; }
+
+	/**
+	 * Reads a PNG and notes all of its colors
+	 *
+	 * This code is more complicated than strictly necessary, but that's because of the API
+	 * being used: the "high-level" interface doesn't provide all the transformations we need,
+	 * so we use the "lower-level" one instead.
+	 * We also use that occasion to only read the PNG one line at a time, since we store all of
+	 * the pixel data in `pixels`, which saves on memory allocations.
+	 */
+	explicit Png(std::filesystem::path const &filePath) : path(filePath), colors() {
+		if (file.open(path, std::ios_base::in | std::ios_base::binary) == nullptr) {
+			fatal("Failed to open input image (\"%s\"): %s", path.c_str(), strerror(errno));
+		}
+
+		options.verbosePrint("Opened input file\n");
+
+		std::array<unsigned char, 8> pngHeader;
+
+		if (file.sgetn(reinterpret_cast<char *>(pngHeader.data()), pngHeader.size())
+		        != static_cast<std::streamsize>(pngHeader.size()) // Not enough bytes?
+		    || png_sig_cmp(pngHeader.data(), 0, pngHeader.size()) != 0) {
+			fatal("Input file (\"%s\") is not a PNG image!", path.c_str());
+		}
+
+		options.verbosePrint("PNG header signature is OK\n");
+
+		png = png_create_read_struct(PNG_LIBPNG_VER_STRING, (png_voidp)this, handleError,
+		                             handleWarning);
+		if (!png) {
+			fatal("Failed to allocate PNG structure: %s", strerror(errno));
+		}
+
+		info = png_create_info_struct(png);
+		if (!info) {
+			png_destroy_read_struct(&png, nullptr, nullptr);
```

Comment:
The `~Png` destructor calls this anyway, and destructors are called at `exit`, so this can go.
```suggestion
```

Reply:
Ah, never mind, the destructor is not called if the constructor hasn't completed. But then, I think the constructor here should be simplified, and call some kind of init method to set up the Png. Since it's not that unusual for reading stuctured data from a file to fail.

Reply:
What would be the point, if the `Png` cannot be used until the "setup" function has been called?

Reply:
That makes sense.